firefox-ios/Storage/Rust/RustRemoteTabs.swift

```diff
@@ -175,140 +177,162 @@ public class RustRemoteTabs {
+        self.getUnsentCommandsByDeviceId(deviceId: deviceId) { commands in
+            let urls = commands.map { item in
+                switch item.command {
+                case .closeTab(let url):
+                    return url
                 }
             }
+            completion(urls)
         }
     }
 
-    private func getUnsentCommandsByDeviceId(deviceId: String, completion: @escaping ([PendingCommand]) -> Void) {
-        guard let commandQueue = self.commandQueue else {
+    public func setPendingCommandsSent(deviceId: String, unsentCommandUrls: [String] = [String]()) {
+        guard let tabsCommandQueue = self.tabsCommandQueue else {
             let err = TabsApiError.UnexpectedTabsError(reason: "Command queue is not initialized") as MaybeErrorType
             self.logger.log(err.description,
                             level: .warning,
                             category: .tabs)
-            completion([PendingCommand]())
             return
         }
 
-        commandQueue.getUnsentCommands { result in
-            switch result {
-            case .success(let commands):
-                let filteredCommands = commands.filter { $0.deviceId == deviceId }
-                completion(filteredCommands)
-            case .failure(let error):
-                self.logger.log("Failed to get unsent commands: \(String(describing: error))",
-                                level: .warning,
-                                category: .tabs)
-                completion([PendingCommand]())
+        self.getUnsentCommandsByDeviceId(deviceId: deviceId) { commands in
+            let sentCommands = self.getSentCommands(unsentCommandUrls: unsentCommandUrls, commands: commands)
+
+            // mark the commands we know to be successfully sent as sent so we don't attempt to send the
+            // commands again
+            tabsCommandQueue.setPendingCommandsSent(deviceId: deviceId, commands: sentCommands) { errors in
+                if errors.isEmpty {
+                    return
+                } else {
+                    let errHeader = "Failed to set some pending commands as sent:\n"
+                    let errMessages: String = errors.reduce(errHeader, { result, err in
+                        result.appending("  - \(err.localizedDescription)\n")
+                    })
+                    self.logger.log(errMessages,
+                                    level: .warning,
+                                    category: .tabs)
+                }
             }
         }
     }
 
-    private func handleCommandQueueErrors(result: Result<Bool, any Error>) {
-        switch result {
-        case .success(let didSucceed):
-            if !didSucceed {
-                self.logger.log("Unexpected error while attempting to update command queue",
+    func getSentCommands(unsentCommandUrls: [String], commands: [PendingCommand]) -> [PendingCommand] {
```

Comment:
This can be a bare function, since it doesn't depend on `self`, and let's give it a name like `filterSentCommands`.

There are a couple too many `get*Commands` methods on this class already, and those all do I/O.